core/src/replicator.rs

```diff
@@ -599,6 +600,126 @@ impl Replicator {
             "Couldn't get blockhash or slot",
         ))?
     }
+
+    /// Ask a replicator to populate a given blocktree with its segment.
+    /// Return the slot at the start of the replicator's segment
+    ///
+    /// It is recommended to use a temporary blocktree for this since the download will not verify
+    /// blobs received and might impact the chaining of blobs across slots
+    pub fn download_from_replicator(
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        replicator_info: &ContactInfo,
+        blocktree: &Arc<Blocktree>,
+    ) -> Result<(u64)> {
+        // Create a client which downloads from the replicator and see that it
+        // can respond with blobs.
+        let start_slot = Self::get_replicator_segment_slot(replicator_info.storage_addr);
+        info!("Replicator download: start at {}", start_slot);
+
+        let exit = Arc::new(AtomicBool::new(false));
+        let (s_reader, r_reader) = channel();
+        let repair_socket = Arc::new(UdpSocket::bind("0.0.0.0:0").unwrap());
```

Comment:
Is port=0 going to have problems with NATs?

Reply:
addr:port 0 just means bind to "any". So it should be fine since I think. I'm not sure about other platforms though. @mvines ?

Reply:
As long as there's an outbound packet from `repair_socket` first, most NATs I've seen should be able to map any response packets back to the original private IP/port.

Reply:
yeah there should be an outbound packet first.

Reply:
Ok. We just probably want replicators to work behind NATs without port forwarding.

Reply:
> Ok. We just probably want replicators to work behind NATs without port forwarding.

Yeah for sure.

Reply:
Yeah for sure. This PR _should_ be okay but I am slightly worried about them being able to receive the ledger in the first place. I think it should be okay since each socket sends an outbound packet first.